### DISCUSSION TITLE
OneAPI usability

### DISCUSSION THREAD

**Discussion by andrii0lomakin** [beehive-lab/TornadoVM#529] [Q&A]
Good day, TornadoVM developers.
We are considering using TornadoVM in one of our projects, and we are analyzing different backends supported.
One of the primary requirements - support virtually all devices on which Java can run.

From this point of view OpenCL looks very appealing, but from other side concept of oneAPI and its supprot of spir-v and its extensions looks very interesting. 

There are several question that we want to resovle first:
1. Support of libraries, we are planning to use clBlast and  cuBLAS libraries (we completely aware that right now thire is no integration with TornadoVM, we are thinking about providing one), but what about oneAP,I is it possible to use oneAPI libraries with level zero backend interchangably ?
2. Usability of installation for end users:
 - Is it correct that Windows installer of level zero does not exist ?
 - Are there level zero docker containers ?

We are going to conduct research ourselves , but we appreaciate if TornadoVM developers will share their expereince in context of those questions.

**Comment by jjfumero** [comment]
Hi @andrii0lomakin , thanks for your interest. 

> Support of libraries, we are planning to use clBlast and cuBLAS libraries (we completely aware that right now thire is no integration with TornadoVM, we are thinking about providing one), but what about oneAP,I is it possible to use oneAPI libraries with level zero backend interchangably ?

I proposed this change for TornadoVM back in JVMLS 2023:  https://youtu.be/VTzGlnv6nuA?feature=shared&t=1673
What we called the hybrid API. Is this what you mean? 
We are internally working towards that. 

For Level Zero interoperability, you might ask Intel oneAPI folks for a more detailed answer. From my view, it really depends on how the kernels are represented. If they are compiled to SPIR-V, then you may be able to use the Intel Level Zero API.  Level Zero is just one of the many software components and  layers of oneAPI. 

> Is it correct that Windows installer of level zero does not exist ? 

No, it is not correct. Level Zero works for Windows too: https://github.com/oneapi-src/level-zero/issues/56 

TornadoVM runs also on Windows with SPIR-V and level Zero. 

> Are there level zero docker containers ? 

oneAPI docker images should work with Level Zero too.

**Reply by andrii0lomakin** [reply]
>No, it is not correct. Level Zero works for Windows too: https://github.com/oneapi-src/level-zero/issues/56

I am afraid that is what I meant, actually. No user will build it from the source code. 

P.S. Will answer the rest of the questions tomorrow.

**Comment by andrii0lomakin** [comment]
>I proposed this change for TornadoVM back in JVMLS 2023: https://youtu.be/VTzGlnv6nuA?feature=shared&t=1673

Yes, I have seen this talk and was wondering why this feature is absent. 

I plan to implement this feature. 
I will start working on it once I finish with https://github.com/beehive-lab/TornadoVM/issues/526. 
If, of course,  you do not implement it first.
 
>oneAPI docker images should work with Level Zero, too.

Do you mean you tested them, or is that your assumption?

>For Level Zero interoperability, you might ask Intel oneAPI folks for a more detailed answer. 

Yeah sure will do.
